db/maintainer.js

```diff
@@ -13,23 +13,101 @@ class MaintainerDbController {
     return pkgs.map(({ _id: id, ...rest }) => ({ id, ...rest }))
   }
 
-  async getRevenue ({ maintainerId }) {
-    const packages = await this.db.collection('packages').find({
-      maintainers: { $elemMatch: { maintainerId } }
-    }).toArray()
-    return packages.reduce((totalRevenue, pkg) => {
-      const { revenuePercent } = pkg.maintainers.find((maintainer) => maintainer.maintainerId === maintainerId)
-      return totalRevenue + (pkg.totalRevenue * (revenuePercent / 100))
-    }, 0)
-  }
-
   async updateIlpPointer ({ userId, ilpPointer }) {
     return this.db.collection('users').updateOne({
       _id: ObjectId(userId)
     }, {
       $set: { 'payoutInfo.ilpPointer': ilpPointer }
     })
   }
+
+  // @returns an array of [{ _id: <packageId>, payout: <payoutInMS> }]
+  async getPendingPayout ({ maintainerId }) {
+    return this.db.collection('packages').aggregate([
+      {
+        $match: {
+          'maintainers.userId': {
+            $eq: maintainerId
+          }
+        }
+      }, {
+        $match: {
+          $or: [
+            {
+              adRevenue: {
+                $ne: null
+              }
+            }, {
+              donationRevenue: {
+                $ne: null
+              }
+            }
+          ]
+        }
+      }, {
+        $project: {
+          adRevenue: {
+            $filter: {
+              input: '$adRevenue',
+              as: 'adRev',
+              cond: {
+                $ne: [
+                  '$$adRev.paid', true
+                ]
+              }
+            }
+          },
+          donationRevenue: {
+            $filter: {
+              input: '$donationRevenue',
+              as: 'donRev',
+              cond: {
+                $ne: [
+                  '$$donRev.paid', true
+                ]
+              }
+            }
+          },
+          maintainers: 1
+        }
+      }, {
```

Comment:
I think this might even be possible to stuff into the `$match` phase so that by the time we get to the `$project` phase (which I think is still necessary), we are already only dealing with packages that have unpaid revenues: https://docs.mongodb.com/manual/tutorial/query-arrays/#query-an-array-with-compound-filter-conditions-on-the-array-elements

Any ways to make the `$match` more specific (as early in the pipeline as possible) means the pipeline will be faster